EZP-27562: Allow square brackets and percent character in RichText links

adamwojs opened the pull request:
> JIRA: https://jira.ez.no/browse/EZP-27562
> Alternative to #2044

## Description

This PR contains patch with alternative approach to solve EZP-27562. In general, the problem is that `anyURI` by definition does not accept `[`,  `]`, `%` characters which are in common use in the URLs e.g. Google Maps. 

I've changed datatype of `href` attribute from `anyURI` to `string` and added regular expression to validate URL format.  Used regex is found by @glye and extended to accept `ezlocation` `ezcontent`, `ezurl` and `#fragment` urls

![zrzut ekranu 2017-09-25 o 14 59 33](https://user-images.githubusercontent.com/211967/30809516-460b4f86-a202-11e7-8249-5627eac0b65c.png)

alongosz commented on eZ/Publish/Core/FieldType/RichText/Resources/schemas/docbook/docbook.iso.sch.xsl:
```diff
@@ -1,82 +1,311 @@
+                contains(@*[name()='xlink:href'], ' ') or
+                contains(@*[name()='xlink:href'], '&#x09;') or
+                contains(@*[name()='xlink:href'], '&#xA;') or
+                contains(@*[name()='xlink:href'], '&#xD;')
```
Note on an implementation itself: I'm not a fan of blacklisting (prefer whitelisting), but since my PR about javascript in hrefs is based also on whitelisting, then I don't object :)
Are you sure this covers every case of possible injection, btw? Not a XSS expert myself ;)

adamwojs commented on eZ/Publish/Core/FieldType/RichText/XSLTProcessorExtension.php:
```diff
@@ -0,0 +1,22 @@
+     * @param string $value
+     * @return bool
+     */
+    public static function isValidUrl($value)
```
Note to reviewer: This method must be public if it is to be accessible from XSL

andrerom replied:
yep, so `XSLTProcessorFuntions` would make that more clear :)

adamwojs replied:
Done.

adamwojs commented on eZ/Publish/Core/FieldType/RichText/XSLTProcessorExtension.php:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Helper functions available in XSL documents.
+ */
+class XSLTProcessorExtension
```
Note to reviewer: Maybe this class should be called "XSLTProcessorFuntions" or in another way ?

andrerom replied:
yes, that would be slightly better. but no strong opinion on it here.

adamwojs commented on eZ/Publish/Core/FieldType/RichText/Resources/schemas/docbook/docbook.iso.sch.xsl:
```diff
@@ -79,4 +79,7 @@
 
 		<!--ASSERT -->
-<axsl:choose><axsl:when test="not(contains(@*[name()='xlink:href'], 'javascript:') or contains(@*[name()='xlink:href'], 'vbscript:'))"/><axsl:otherwise><svrl:failed-assert xmlns:svrl="http://purl.oclc.org/dsdl/svrl" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:schold="http://www.ascc.net/xml/schematron" test="not(contains(@*[name()='xlink:href'], 'javascript:') or contains(@*[name()='xlink:h ...
+<axsl:choose><axsl:when test="not(contains(@*[name()='xlink:href'], 'javascript:') or contains(@*[name()='xlink:href'], 'vbscript:'))"/><axsl:otherwise><svrl:failed-assert xmlns:svrl="http://purl.oclc.org/dsdl/svrl" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:schold="http://www.ascc.net/xml/schematron" test="not(contains(@*[name()='xlink:href'], 'javascript:') or contains(@*[name()='xlink:h ...
```
Note to reviewer: Those two validators overlaps. Should we remove the first one as more specific?

andrerom replied:
referring to the `axsl:when`  vs the `svrl:failed-assert` on this line?

Or in relevance to line below?

adamwojs replied:
I'm referring to those selected lines https://github.com/adamwojs/ezpublish-kernel/blob/bedace3e674301cb60dff863b94528ee2cc7620c/eZ/Publish/Core/FieldType/RichText/Resources/schemas/docbook/docbook.iso.sch#L255-L256  

_I should be more precise, sorry for that_

andrerom replied:
Ok, maybe @alongosz has input on this?

adamwojs replied:
The only one side effect of using both of them at once double validation error:
![zrzut ekranu 2017-10-02 o 16 37 48](https://user-images.githubusercontent.com/211967/31082831-3bea7084-a790-11e7-9ad7-a701c7a0eec3.png)

andrerom replied:
ah, we don't want that.. could you find a agreement with @alongosz ? @pspanja is back from vacation so might be possible to reach as well ;)

alongosz replied:
AFAICS full error msg of your validator here is `/section/para[2]/link: The value must be a valid URL`, right? If that's the case, then it's fine to remove my validator, it's ugly :-)
BTW. How does it look like in the case of multiple invalid links in multiple paragraphs?

But please see this general remark I posted some time ago:
> General remark: `docbook.iso.sch.xsl` is generated from `docbook.iso.sch` and the latter one should be changed (Sorry for the inconvenience!). 
See [README](https://github.com/ezsystems/ezpublish-kernel/blob/v6.11.2/eZ/Publish/Core/FieldType/RichText/Resources/schemas/docbook/Readme.rst#ez-publish-docbook-variant) to find out how to generate it.

So, in short, it cannot be delivered in the current form.
Two directions:
1. Generate this XSL using schematron file as described above.
2. If it's impossible to produce the samle xsl using schematron, then maybe we should abandon schematron and change file suffix from `.sch.xsl` to just `.xsl`

alongosz replied:
Please disregard part about schematron, I didn't notice source file .sch is also a part of this PR. Now it's fine.

glye commented on eZ/Publish/Core/FieldType/RichText/XSLTProcessorExtension.php:
```diff
@@ -0,0 +1,22 @@
+ */
+class XSLTProcessorExtension
+{
+    const URL_PATTERN = '%^((ezlocation|ezurl|ezcontent):\/\/\d+(#[^\s]+)?)|(#[^\s]+)|(?:(?:https?|ftp)://)(?:\S+(?::\S*)?@|\d{1,3}(?:\.\d{1,3}){3}|(?:(?:[a-z\d\x{00a1}-\x{ffff}]+-?)*[a-z\d\x{00a1}-\x{ffff}]+)(?:\.(?:[a-z\d\x{00a1}-\x{ffff}]+-?)*[a-z\d\x{00a1}-\x{ffff}]+)*(?:\.[a-z\x{00a1}-\x{ffff}]{2,6}))(?::\d+)?(?:[^\s]*)?$%iu';
```
If this is based on what I used in https://github.com/ezsystems/ezpublish-kernel-ee/pull/161/files#diff-981d67ab6f6be0f1d1ff989637c9e357R82 I think it would be good to reference where it comes from, like I did there, just to give them credit.

adamwojs replied:
I've added note about regex author in docblock on `isValidUrl` method

adamwojs commented:
> I assume you considered the alternative: Keep using anyURI but %-encode as needed. Why did you prefer the current solution?

@glye Yes, we took this into consideration in the first attempt to solve this problem. See https://github.com/ezsystems/ezpublish-kernel/pull/2044

adamwojs commented:
PR updated, changelog:

* Added Copyright in XSLTProcessorFunctions.php
* Added Unit tests for `XSLTProcessorFunctions`
* Added support for `ftps://` protocol
* Removed old overlapping validator for `javascript:` and `vbscript:`
